tests/sentry/integrations/gitlab/test_repository.py

```diff
@@ -0,0 +1,131 @@
+            }
+        )
+        self.integration.add_organization(self.organization, self.user, identity.id)
+        self.integration.get_provider().setup()
+
+        self.default_repository_config = {
+            'path_with_namespace': 'getsentry/example-repo',
+            'name_with_namespace': 'Get Sentry / Example Repo',
+            'path': 'example-repo',
+            'id': 123,
+            'web_url': 'https://example.gitlab.com/my-group/projects/example-repo',
+        }
+
+    @fixture
+    def provider(self):
+        return GitlabRepositoryProvider()
+
+    def create_repository(self, repository_config, integration_id, organization_slug=None):
+        repo_id = repository_config['id']
+        responses.add(
+            responses.GET,
+            u'https://example.gitlab.com/api/v4/projects/%s' % repo_id,
+            json=repository_config
+        )
+
+        with self.feature({'organizations:internal-catchall': True, 'organizations:repos': True}):
+            response = self.client.post(
+                path=reverse(
+                    'sentry-api-0-organization-repositories',
+                    args=[organization_slug or self.organization.slug]
+                ),
+                data={
+                    'provider': self.provider_name,
+                    'installation': integration_id,
+                    'identifier': repo_id,
+                }
+            )
+        return response
+
+    def assert_repository(self, repository_config, organization_id=None):
+        domain_name = self.integration.metadata['domain_name']
+        repo = Repository.objects.get(
+            organization_id=organization_id or self.organization.id,
+            provider=self.provider_name,
+            external_id='%s:example-repo' % (domain_name,)
+        )
+        assert repo.name == repository_config['name_with_namespace']
+        assert repo.url == repository_config['web_url']
+        assert repo.integration_id == self.integration.id
+        assert repo.config == {
+            'instance': domain_name,
+            'repo_id': repository_config['id'],
+            'path': repository_config['path_with_namespace'],
+        }
+
+    @responses.activate
+    def test_create_repository(self):
+        response = self.create_repository(self.default_repository_config, self.integration.id)
+        assert response.status_code == 201
+        self.assert_repository(self.default_repository_config)
+
+    def test_create_repository_null_installation_id(self):
+        response = self.create_repository(self.default_repository_config, None)
+        assert response.status_code == 500
+
+    def test_create_repository_integration_does_not_exist(self):
+        integration_id = self.integration.id
+        self.integration.delete()
+
+        response = self.create_repository(self.default_repository_config, integration_id)
+        assert response.status_code == 500  # TODO(lb): shouldn't this result in a 404?
```

Comment:
Couple of error responses I was curious about. This one happens when an integration is not found. Would a 404 be better here?

Reply:
Yes a 404 indicates that the end user was looking for something that didn't exist. A 500 usually means the customer did something valid but we failed to handle the request.

Reply:
I think this is just default behavior. I'll see what I need to do to change it.